### PR TITLE
Implement `__array__` for `qasm2._DefinedGate`

### LINKED ISSUE #12100: `XXPlusYYGate.to_matrix()` fails after qasm2 roundtrip
### Environment

- **Qiskit version**: 1.0.2
- **Python version**: 3.11.0
- **Operating system**: Linux

### What is happening?

After a qasm2 round-trip, a `XXPlusYYGate`'s `to_matrix()` method no longer works.

~~This problem seems somewhat special to `XXPlusYYGate`; it does not exist for most other two-qubit gates we have tried.~~

I've been able to reproduce this problem as well for `CSGate`, but not `CXGate` or `CZGate`.

### How can we reproduce the issue?

```python
from qiskit import qasm2, QuantumCircuit
from qiskit.circuit.library import XXPlusYYGate

# Build a circuit with a single XXPlusYYGate
qc = QuantumCircuit(2)
qc.append(XXPlusYYGate(0.1), [0, 1])

# Perform a qasm2 round trip
program = qasm2.dumps(qc)
qc2 = QuantumCircuit.from_qasm_str(program)

# Call to_matrix on original circuit instructions (works)
for inst in qc.data:
    inst.operation.to_matrix()

# Call to_matrix on circuit instructions after serialization (fails)
for inst in qc2.data:
    inst.operation.to_matrix()
```

throws the exception

```
File ~/serverless/.direnv/python-3.11.0/lib/python3.11/site-packages/qiskit/circuit/gate.py:63, in Gate.to_matrix(self)
     61 if hasattr(self, "__array__"):
     62     return self.__array__(dtype=complex)
---> 63 raise CircuitError(f"to_matrix not defined for this {type(self)}")

CircuitError: "to_matrix not defined for this <class 'qiskit.qasm2.parse._DefinedGate'>"
```

### What should happen?

it should work the same as the circuit that did not undergo a round trip.

### Any suggestions?

_No response_

### PR DESCRIPTION
**Pull request by jakelishman**
### Summary

Gates defined from OpenQASM 2 should have a well-defined matrix form (up to global phase) whenever all the constituent parts of the definition do. This manually makes such a matrix available.

### Details and comments

Fix #12100.

### CODE REVIEW

File: qiskit/qasm2/parse.py [hunk 1]
```diff
@@ -318,6 +321,9 @@ def _define(self):
                 raise ValueError(f"received invalid bytecode to build gate: {op}")
         self._definition = qc
 
+    def __array__(self, dtype=None):
```
Comment (mtreinish) [inline qiskit/qasm2/parse.py]: This will need a copy kwarg for numpy 2.0 support right?

Comment (jakelishman) [reply qiskit/qasm2/parse.py]: Yeah, I wrote this PR before the Numpy 2.0 one

Comment (jakelishman) [reply qiskit/qasm2/parse.py]: Should be done in 88ff0d6a90.

**Review by mtreinish — APPROVED** [review]
This LGTM now, but I'm going to hold off on auto merging it to stable/1.1 for 1.1.0. I think it's a bit out of scope for backport at this point. We can discuss it for 1.1.1 (I think we might want to just hold it for 1.2.0)

**Pull request merged**